### ISSUE TITLE
Loss not decreasing in SWAV pretext task

### ISSUE THREAD

**Issue by trancenoid** [KeremTurgutlu/self_supervised#62]
Discussion link is not working so I posted here. I am using SWAV with efficientnet-b4, adam opt. and lr = 0.311. My dataset is small (around 1200 images ) 370x300 in dimension and is black and white (1-channel). For some reason the train and valid loss doesn't decrease beyond 1st epoch.

What I have already tried :
1. Removing normalization
2. Increasing / decreasing batch size
3. Setting pretrained = False 
4. Changing projection_size = 128

None gave any improvement

Here is the code to load data and prepare the model

```

from self_supervised.vision.swav import *
from self_supervised.layers import create_encoder

item_tfms = [Resize((376, 300), method='squish')]
batch_tfms = [Normalize.from_stats(*imagenet_stats)]
bs = 8

dls = ImageDataLoaders.from_df(all_data, valid_col = 'is_valid', item_tfms = item_tfms, fn_col = 'img',
                               batch_tfms = None, bs = bs, label_col = 'label')



encoder = create_encoder("tf_efficientnet_b4_ns", n_in=3, pretrained=True) 


model = create_swav_model(encoder, hidden_size=2048, projection_size=512)
aug_pipelines = get_swav_aug_pipelines(num_crops=[2,6],
                                       crop_sizes=[128,96], 
                                       min_scales=[0.25,0.05],
                                       max_scales=[1.0,0.3])

learn = Learner(dls, model, cbs=[SWAV(aug_pipelines=aug_pipelines, crop_assgn_ids=[0,1], K=bs*2**6, queue_start_pct=0.5)])
```

The data cannot be shared but is of similar nature as medical scans.

**Comment by trancenoid** [comment]
Thanks for the quick reply and apologies for limited info on the problem,
1. I will have to search for a similar data, but as you described in (2), there is high inter class variance (samples from same class appear in varied forms) and low inter class variance (sample in different class looks similar on the overall scale). You can imagine it like finding anomalies (scratch marks, dust etc) in integrated circuits, given its top view, and each image is perfectly aligned.
2. I am playing around with augmentations. Please correct me if I am wrong here, the augmentation should not make the image so twisted that it is impossible even for human to judge the original image, and still retain the features I want in the end (small regions, scratches, dust). So I dropped blur, and only have small amount of crop by reducing the min and max scale.
3. Thanks for the link, I will go through the relevant literature.
4. Not sure what and how to check that, can you shed more light here.

Currently this is my model code:
```

from fastai.vision.all import *
from self_supervised.vision.swav import *
from self_supervised.layers import create_encoder
size = 320
batch_tfms = [] 
bs = 12

dls = ImageDataLoaders.from_df(df, item_tfms = item_tfms, valid_col = 'is_valid', fn_col = "fname",
                                   batch_tfms=batch_tfms, bs = bs, label_col = "dum")



encoder = create_encoder("xresnet34", n_in=3, pretrained=True) # a fastai encoder/


model = create_swav_model(encoder, hidden_size=2048, projection_size=64)
aug_pipelines = get_swav_aug_pipelines(num_crops=[2,6],
                                       crop_sizes=[size,int(0.9*(size))], 
                                       min_scales=[0.75,0.85],
                                       max_scales=[1.0,0.9], blur = False)


learn = Learner(dls, model, cbs=[SWAV(aug_pipelines=aug_pipelines, crop_assgn_ids=[0,1], K=bs*2**6, queue_start_pct=0.5), ReduceLROnPlateau()])
```
I also need suggestions on projection_size and batch size. I cannot go below 320x320 since the defects will be barely visible. I have single v100 16GB GPU which I am using to train.

**Comment by KeremTurgutlu** [comment]
> I am playing around with augmentations. Please correct me if I am wrong here, the augmentation should not make the image so twisted that it is impossible even for human to judge the original image, and still retain the features I want in the end (small regions, scratches, dust). So I dropped blur, and only have small amount of crop by reducing the min and max scale.

This paper: [What Makes for Good Views for Contrastive Learning?](https://arxiv.org/pdf/2005.10243.pdf) can help build some intuition around data augmentations for self-supervised learning. I think the general idea is to remove noise as much as you can yet still to keep the task relevant signals.

> Not sure what and how to check that, can you shed more light here?

Check the embeddings after some training. Perhaps model outputs the same embedding for any given image as a trivial solution, this is referred as representation collapse in literature. Outputting same representation for any given image.

Additional to data augmentations, you can also try different learning rate, learning rate scheduling, different K (which is the # of prototypes/clusters).
